### ISSUE TITLE
retrieval augmented ChatGPT

### ISSUE THREAD

**Issue by bruffridge** [nasa-petal/discord_bot#2]
Use retrieval methods (eg. vector embeddings similarity, or BM25) to find information from journal articles to provide useful context to improve the outputs from ChatGPT for the Discover and Abstract steps of BDA mode.

Use of existing API based search capabilities are preferred rather than building one from scratch if they work well. Available search APIs are listed here: https://github.com/nasa-petal/PeTaL/wiki/BID-resources

Brief description of retrieval augmentation:
https://youtu.be/bZQun8Y4L2A?t=1974

Embeddings for all the papers in Arxiv.org
https://alex.macrocosm.so/download

Medium article showing how to use some existing tools that can do something similar:
https://medium.com/a-academic-librarians-thoughts-on-open-access/using-large-language-models-like-gpt-to-do-q-a-over-papers-ii-using-perplexity-ai-15684629f02b

https://github.com/openai/openai-cookbook/blob/main/examples/Question_answering_using_embeddings.ipynb

This may be helpful:
https://twitter.com/danshipper/status/1615901860786749440
He used this to load his reference library:
https://github.com/jerryjliu/gpt_index

This is another example that may be useful:
https://every.to/superorganizers/i-trained-a-gpt-3-chatbot-on-every-episode-of-my-favorite-podcast

**Comment by bruffridge** [comment]
Three uses for retrieval augmentation, to help improve `bdamode`.

1. Discover additional relevant biological models that ChatGPT didn't generate on its own.
2. Find accurate and relevant sources to include for each biological strategy, rather than relying on ChatGPT to generate them (since it has been shown to get this wrong sometimes).
3. Provide context to help ChatGPT generate more accurate and detailed descriptions of how the biological strategies work.

**Comment by bruffridge** [comment]
from @hschilling 

For Better Answers, Generate Reference Text
If you want a model to answer questions correctly, then [enriching the input with reference text retrieved from the web](https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Finfo.deeplearning.ai%2Fe3t%2FCtc%2FLX%2B113%2FcJhC404%2FVVyHRw1qYcSbW4H36J56P47wkW8ZpTlp4-6dy4N84392_3q3n5V1-WJV7CgCydVBRTsC1ftZxJW3kmryT7HKhQvW7rBfzv4DQLKLW1j9Y3B2_Jh22W3m3H7g8Xv1V2W10bQm26xf5JzW43T5V566148RW7dj_N68DT-kkW9hDtxC7Bq9RZW5WHF7M2fJzccW3Y3dbV95-BNxW3YCt5-7gfxG-W3P467g5vrFVGW4z061p2G6mB7W3HKxXS1z4Ql9W80NZbh1yg59wW3YwRxv4V3sX1W7cybYz82-gVrW7SbtRp2VNSqQVMbf6t7yFkFW36f61&data=05%7C01%7Cbruffridge%40nasa.gov%7C79deff05461f452e720d08db4c156b55%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C638187427095979477%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=JTyZ8r2PyaOwIGgXhPp6EQBD2yheb24agWdNJmhwE7Q%3D&reserved=0) is a reliable way to [increase the accuracy of its output](https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Finfo.deeplearning.ai%2Fe3t%2FCtc%2FLX%2B113%2FcJhC404%2FVVyHRw1qYcSbW4H36J56P47wkW8ZpTlp4-6dy4N84392_3q3n5V1-WJV7CgNb9N7grbVtHXjB2W8T7ZR-28rSpYW1GPTLL3zXxXRW7N76D838PZs2W8VG4zf7TG1SGVhZfwN6D0xVXW54rc5S8FPR-nW5SkW5J6BlnNfW4CNwL_7VV_pjW5V4lRq6-VGVmW5TKyvm4ZlDKYW32lJFv30j9tWW7z4VSl4TWn-8W53Zfjb6_yYN3W3NxXWD8VdWwGW2tCqwP1j5x7GW2hxryr7X0YG7W7pJ4ZB68F6KZW8vgWng8pGjBTW8r-PgN25LB4n32FL1&data=05%7C01%7Cbruffridge%40nasa.gov%7C79deff05461f452e720d08db4c156b55%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C638187427095979477%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=TE5TB1gYeKkfsnBDn9OowM5BLxia7xMMRW6UPVCmYEw%3D&reserved=0). But the web isn’t necessarily the best source of reference text.
What's new: Wenhao Yu at University of Notre Dame and colleagues at Microsoft and University of Southern California used a pretrained language model to generate reference text. They fed that material, along with a question, to a second pretrained language model that [answered](https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Finfo.deeplearning.ai%2Fe3t%2FCtc%2FLX%2B113%2FcJhC404%2FVVyHRw1qYcSbW4H36J56P47wkW8ZpTlp4-6dy4N84392_3q3n5V1-WJV7CgKwBW3Ry5yd2vXR-JW3ZV1HX1jB7DGW42ftsw2300H9W6wmNC_3b6DrBW3ky8z62rlJC6W7R5zCK4RDZBrVrLx3l5Ls5p4W3LnnC73rD2xvVSHN_L16P7G9W93FcxR7Ql5cpW5jYq4v4hzc1zW5KbVRT3JBVhKW8QFYbV5y6ws2W1s8GLd83rw5GVJQD7w80n6ZpW8yySKV6vBhZkW81g0zv7kxBr2W7ybFS313SzD5W2MGSRK3Nzr8HW1JVyq75B9pF336nv1&data=05%7C01%7Cbruffridge%40nasa.gov%7C79deff05461f452e720d08db4c156b55%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C638187427095979477%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=vkepBVRVBGLm5%2BpnN4vfEP4jEFvd9aQIWkit68L4%2FbE%3D&reserved=0) more accurately than a comparable model that was able to retrieve relevant text from the web.
Key insight: Given a question, documents retrieved from the web, even if they’re relevant, often contain information that doesn’t help to answer it. For instance, considering the question “How tall is Mount Everest?,” the Wikipedia page on Mount Everest contains the answer but also a lot of confusing information such as elevations attained in various attempts to reach the summit and irrelevant information that might distract the model. A language model pretrained on web pages can generate a document that draws on the web but focuses on the question at hand. When fed to a separate language model along with the question, this model-generated reference text can make it easier for that model to answer questions correctly.
How it works: The authors used a pretrained [InstructGPT](https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Finfo.deeplearning.ai%2Fe3t%2FCtc%2FLX%2B113%2FcJhC404%2FVVyHRw1qYcSbW4H36J56P47wkW8ZpTlp4-6dy4N84392_3q3n5V1-WJV7CgPDXW2tYR4y2qxBTvW6LtKC669rSfbW4V7Vxc7hCVcfVsKQ4Z7ps8rpW6n6B0l3BN6JRW3vR_Z-7ysWSvW75HWjP7Qh0lqW5YN39J7f05KXVD1Mfd12nZRhW3Y_gTp3bYmPcV8-k-D30RxdGW89J3dX3KCKgKW91lgx56LMPQcN12dgGbWXGTSW94Hd-r7FmMjXW65t6X15T64nRW8lBZVL2rWC2yW8FWt5n5-VsnhW3cM-t18RlP5gW7-MlST7qF53M31gV1&data=05%7C01%7Cbruffridge%40nasa.gov%7C79deff05461f452e720d08db4c156b55%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C638187427095979477%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=YGcQjziGVZR78RgGj8R4sj0psS6nO8U%2Bw%2BRHkOJ4KGk%3D&reserved=0) (175 billion parameters) to generate reference text related to questions in trivia question-answer datasets such as [TriviaQA](https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Finfo.deeplearning.ai%2Fe3t%2FCtc%2FLX%2B113%2FcJhC404%2FVVyHRw1qYcSbW4H36J56P47wkW8ZpTlp4-6dy4N84393h3q3npV1-WJV7CgXCzW1SXlbY5QCzvTN88PZDMslLVYN1FLZhzw8Lk3W5khtDg7RmsxXW8g8_6B6n0J5pW5c9L8F40DCCqW6t5MP137TSbbW7b5NWV2ZW89MW6gkWyc2cv086W6thPfV4Q6fs-W65HTP63CgwMtVkL4Jf956YWBW1WqHz84Cln6WW481PwT26DVf3W74Mdz42X3HB1W88YHkV78d3XxW6yrh4N3lL6-_W1YtH6Z1kYhxjW1npHYV6Vz566N5QMKKjm-C4bW4p5ryR5gwpBYW7QD4TK4Y1dD2333-1&data=05%7C01%7Cbruffridge%40nasa.gov%7C79deff05461f452e720d08db4c156b55%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C638187427095979477%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=65FQ%2FqWNU1jzFPB6cpd0U8KyCfKbbRvJuDqDeSI%2FAjU%3D&reserved=0). They generated answers using [FiD](https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Finfo.deeplearning.ai%2Fe3t%2FCtc%2FLX%2B113%2FcJhC404%2FVVyHRw1qYcSbW4H36J56P47wkW8ZpTlp4-6dy4N84392_3q3n5V1-WJV7CgT5xV4xXBZ8vH-MlW3714qK2wTN4WW74mmZ92_b_0kW3d9SpZ4gghYRN477xjMLfZChW70J_5q8CvxmbW22DRtY8sSgC7N6v-GDYvRy7BW2vNq_M7xjv6RW5FXwdX5r_knjVFzQH42g8skxW90___88k4jqwW2_-Y3n3rHylZW44G9Gd2s_l8YN8HtMyYw558yW83wGjr4Q0MByVxdzST1lN_qmN4MdplK9Sg-ZW20pdMS2WPjZ-W4jVDVM3ZCNnd2Cg1&data=05%7C01%7Cbruffridge%40nasa.gov%7C79deff05461f452e720d08db4c156b55%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C638187427095979477%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=vZbI9Qiwmmq9bn9xs93xbmh2ox5NBLIUGRcJUS4MtFM%3D&reserved=0) (3 billion parameters), which they had fine-tuned on the dataset plus the reference text. (A given question may have more than one valid answer.)
InstructGPT generated reference text for each question in the dataset based upon a prompt such as, “Generate a background document to answer the given question,” followed by the question.
The authors embedded each question-reference pair using [GPT-3](https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Finfo.deeplearning.ai%2Fe3t%2FCtc%2FLX%2B113%2FcJhC404%2FVVyHRw1qYcSbW4H36J56P47wkW8ZpTlp4-6dy4N84392_3q3n5V1-WJV7CgJt4W3n5nTy3RvxFPW8Kq8Cx52_9MvW7YgCVY2MDJ1jW2M6s-s9gNgg7W5394qQ17-56vW2tXPqv7BDFY2W8DQdYr1jv9jjW1JWyPl5GMzzhW4P4JNt7cV8M0W5X3lsX6fkkW_W5HSpVT172MXkW8TdpXK7cXCZ7Vf3qCs7bDBJwW4CnqyJ1XJ1H8W1LsFvX53J8n_W2bpgtr7N_1QnW8jXrgy1bMShXMVBYDHK3z_qW2Km1007gTg5LN41FX9gNNZcM3pCH1&data=05%7C01%7Cbruffridge%40nasa.gov%7C79deff05461f452e720d08db4c156b55%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C638187427095979477%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=fggBcb1WUVndOYreucyGVyKBhJc9skVIH4Ac05JdlvI%3D&reserved=0) and clustered the embeddings via k-means. 
At inference, the system randomly selected five question-reference pairs from each cluster — think of them as guide questions and answers. 
For each cluster, given an input question (such as, "What type of music did Mozart compose?") and the question-reference pairs, InstructGPT generated a document — information related to the question.
Given the question and documents, FiD generated an answer. (Valid answers to the Mozart question include, "classical music," "opera," and "ballet.")
Results: The authors evaluated their fine-tuned FiD on [TriviaQA](https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Finfo.deeplearning.ai%2Fe3t%2FCtc%2FLX%2B113%2FcJhC404%2FVVyHRw1qYcSbW4H36J56P47wkW8ZpTlp4-6dy4N84393h3q3npV1-WJV7CgLTrW2HYnbg5xGQPbW93fZVL1rKGSrW8wwvtL6B3b0nN428VC0Z5FD8W89hygr8FlFdSW57fPwL2mQ30LW7h59V_3PwDr1W6wYcX83FWQJMW4D4Xnd66hlz9W3FTDfw7Kl66bW4Fkwrl6Mjzm6W4KlwTC9367s7W64Pb8j1FFwFXW1TVQW18cj942W2p3F3l74-BfrW5bfK0P6QKLl4W4j-DnS1bbP-FW81LQdw7ZMK1zW6FnXlz3QP2B1W6ggtzN7Q5JwRW8dxFyh1_r8tlW6DK5b036RV9J3lNB1&data=05%7C01%7Cbruffridge%40nasa.gov%7C79deff05461f452e720d08db4c156b55%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C638187427095979477%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=TLk9YxLXjODwpZz0aG4SXzI%2FSVpAUbIsfmXfZVou%2F5g%3D&reserved=0) according to the percentage of answers that exactly matched one of a list of correct answers. Provided with generated documents, FiD answered 71.6 percent of the questions correctly compared to 66.3 percent for FiD fine-tuned on TriviaQA and provided with text retrieved from Wikipedia using [DPR](https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Finfo.deeplearning.ai%2Fe3t%2FCtc%2FLX%2B113%2FcJhC404%2FVVyHRw1qYcSbW4H36J56P47wkW8ZpTlp4-6dy4N84392_3q3n5V1-WJV7CgV4cVR4D1F64GWmlW6l0XWm4kzQMzW7CXynL7c75F8W5pS7mr9021MqW3xbPb11jdvFWW7SY0pQ83s7fRVl7fQ96TxGjMW273ly2976ZkQW7wdX6f6BTl5nW4Z5ChW1vvKwcW7Z3XC15njd8rW19-Lsk63S9ZpN1W6X7TNrYkpW2cKT93856JM2W7--2qR25C4yLN6kVkBD9l5ZpW38Fq9r3M1r0PV9S-bC3HPnZZW8SDh_x7bRbBXVr-wS95cYsfz35B61&data=05%7C01%7Cbruffridge%40nasa.gov%7C79deff05461f452e720d08db4c156b55%7C7005d45845be48ae8140d43da96dd17b%7C0%7C0%7C638187427096135688%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=HCCoJ9CJOp6QYu5jcbcp1ZZ0XMkcbJ9Xu67vWTrzrfQ%3D&reserved=0).
Yes, but: The authors’ approach performed best (74.3 percent) when it had access to both Wikipedia and the generated documents. While generated documents may be better than retrieved documents alone, they worked best together.
Why it matters: Good reference text substantially improves a language model’s question-answering ability. While a relevant Wikipedia entry is helpful, a document that’s directly related to the question is better — even if that document is a product of text generation.
We're thinking: Your teachers were right — Wikipedia isn’t the best source
